Sources/NIO/FixedLengthFrameDecoder.swift

```diff
@@ -0,0 +1,57 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+
+///
+/// A decoder that splits the received `ByteBuffer` by a fixed number
+/// of bytes. For example, if you received the following four fragmented packets:
+///
+/// +---+----+------+----+
+/// | A | BC | DEFG | HI |
+/// +---+----+------+----+
+/// 
+/// A `FixedLengthFrameDecoder` will decode them into the
+/// following three packets with the fixed length:
+///
+/// +-----+-----+-----+
+/// | ABC | DEF | GHI |
+/// +-----+-----+-----+
+///
+public final class FixedLengthFrameDecoder: ByteToMessageDecoder {
+
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+
+    public var cumulationBuffer: ByteBuffer?
+
+    private let frameLength: Int
+
+    /// Create `FixedLengthFrameDecoder` with a given frame length.
+    ///
+    /// - parameters:
+    ///    - frameLength: The length of a frame.
+    public init(frameLength: Int) {
+        self.frameLength = frameLength
+    }
+
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        guard let slice = buffer.readSlice(length: frameLength) else {
```

Comment:
we should do something here when the handler is removed and `buffer` is non-empty. Options that I can currently think of:

- fire the remaining bytes through the pipeline by implementing `handlerRemoved` as `ctx.fireChannelRead(self.wrapInboundOut(self.buffer)); ctx.fireChannelReadComplete()`
- fire an error that we have some left-over

CC @normanmaurer / @Lukasa

Reply:
Why not split the difference and fire an error that contains the left-over bytes?

Reply:
@Lukasa yes, that's also what I think we should do. Maybe we should have a global 'left over stuff' error type with an associated ByteBuffer?

Reply:
Sounds good to me.